File: backend/app/models/access_audit.py
# Copyright (c) 2026 Laurent Barbe
# Licensed under the Apache License, Version 2.0
"""Read-only contracts for the Admin effective-access audit."""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import Field

from app.models.base import ApiModel
from app.models.pagination import PaginatedResponse
from app.models.user import UiRole

AccessAuditScope = Literal["platform", "rgw_account", "rgw_user", "s3_connection"]
AccessAuditSourceKind = Literal["direct", "group"]
AccessAuditRightCode = Literal[
    "ceph_admin",
    "storage_ops",
    "manager_bucket_compare",
    "manager_bucket_integrity_check",
    "manager_bucket_migration",
    "manager_feature_rules",
    "manager_bucket_purge",
    "private_connection_create",
    "managed_private_connection_provision",
    "browser_advanced_features",
    "account_administrator",
    "portal_manager",
    "portal_user",
    "manager_browser_data_access",
    "rgw_user_access",
    "shared_connection_access",
]


class AccessAuditGrantSource(ApiModel):
    kind: AccessAuditSourceKind
    group_id: Optional[int] = None
    group_name: Optional[str] = None


class AccessAuditRight(ApiModel):
    code: AccessAuditRightCode
    label: str
    sources: list[AccessAuditGrantSource] = Field(default_factory=list)


class AccessAuditPrincipal(ApiModel):
    id: int
    email: str
    full_name: Optional[str] = None
    role: UiRole
    is_active: bool = True


class AccessAuditTarget(ApiModel):
    id: Optional[int] = None
    name: str
    identifier: Optional[str] = None


class AccessAuditRow(ApiModel):
    key: str
    principal: AccessAuditPrincipal
    scope: AccessAuditScope
    target: AccessAuditTarget
    rights: list[AccessAuditRight] = Field(default_factory=list)


class PaginatedAccessAuditResponse(PaginatedResponse):
    items: list[AccessAuditRow]

File: frontend/src/api/accessAudit.ts
/*
 * Copyright (c) 2026 Laurent Barbe
 * Licensed under the Apache License, Version 2.0
 */
import client from "./client";
import type { PaginatedResponse } from "./types";

export type AccessAuditScope = "platform" | "rgw_account" | "rgw_user" | "s3_connection";
export type AccessAuditSourceKind = "direct" | "group";
export type AccessAuditRightCode =
  | "ceph_admin"
  | "storage_ops"
  | "manager_bucket_compare"
  | "manager_bucket_integrity_check"
  | "manager_bucket_migration"
  | "manager_feature_rules"
  | "manager_bucket_purge"
  | "private_connection_create"
  | "managed_private_connection_provision"
  | "browser_advanced_features"
  | "account_administrator"
  | "portal_manager"
  | "portal_user"
  | "manager_browser_data_access"
  | "rgw_user_access"
  | "shared_connection_access";

type AccessAuditGrantSource = {
  kind: AccessAuditSourceKind;
  group_id?: number | null;
  group_name?: string | null;
};

export type AccessAuditRight = {
  code: AccessAuditRightCode;
  label: string;
  sources: AccessAuditGrantSource[];
};

export type AccessAuditRow = {
  key: string;
  principal: {
    id: number;
    email: string;
    full_name?: string | null;
    role: string;
    is_active: boolean;
  };
  scope: AccessAuditScope;
  target: {
    id?: number | null;
    name: string;
    identifier?: string | null;
  };
  rights: AccessAuditRight[];
};

export type AccessAuditSortBy = "user" | "scope" | "target";
export type AccessAuditSortDir = "asc" | "desc";

type AccessAuditQuery = {
  page?: number;
  page_size?: number;
  search?: string;
  scope?: AccessAuditScope;
  right?: AccessAuditRightCode;
  source?: AccessAuditSourceKind;
  user_id?: number;
  target_id?: number;
  sort_by?: AccessAuditSortBy;
  sort_dir?: AccessAuditSortDir;
};

export const ACCESS_AUDIT_SCOPES: Array<{ value: AccessAuditScope; label: string }> = [
  { value: "platform", label: "Platform" },
  { value: "rgw_account", label: "RGW Account" },
  { value: "rgw_user", label: "RGW User" },
  { value: "s3_connection", label: "S3 Connection" },
];

export const ACCESS_AUDIT_RIGHTS: Array<{ value: AccessAuditRightCode; label: string }> = [
  { value: "ceph_admin", label: "Ceph Admin" },
  { value: "storage_ops", label: "Storage Ops" },
  { value: "manager_bucket_compare", label: "Manager · Bucket compare" },
  { value: "manager_bucket_integrity_check", label: "Manager · Integrity check" },
  { value: "manager_bucket_migration", label: "Manager · Bucket migration" },
  { value: "manager_feature_rules", label: "Manager · Feature rules" },
  { value: "manager_bucket_purge", label: "Manager · Bucket purge" },
  { value: "private_connection_create", label: "Create manual private connections" },
  { value: "managed_private_connection_provision", label: "Provision managed private connections" },
  { value: "browser_advanced_features", label: "Browser advanced features" },
  { value: "account_administrator", label: "Account administrator" },
  { value: "portal_manager", label: "Portal manager" },
  { value: "portal_user", label: "Portal user" },
  { value: "manager_browser_data_access", label: "Manager Browser data access" },
  { value: "rgw_user_access", label: "RGW user access" },
  { value: "shared_connection_access", label: "Shared S3 connection access" },
];

export async function listAccessAudit(params: AccessAuditQuery = {}): Promise<PaginatedResponse<AccessAuditRow>> {
  const { data } = await client.get<PaginatedResponse<AccessAuditRow>>("/admin/access-audit", { params });
  return data;
}

export async function downloadAccessAuditCsv(params: Omit<AccessAuditQuery, "page" | "page_size"> = {}): Promise<Blob> {
  const { data } = await client.get<Blob>("/admin/access-audit/export.csv", {
    params,
    responseType: "blob",
  });
  return data;
}
